src/v/utils/xid.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <fmt/core.h>
+
+#include <array>
+#include <cstdint>
+#include <exception>
+
+class iobuf;
+class iobuf_parser;
+/**
+ * Exception indicating that provided XID isn't valid f.e. incorrect length of
+ * invalid characters
+ */
+class invalid_xid final : public std::exception {
+public:
+    explicit invalid_xid(const ss::sstring&);
+    const char* what() const noexcept final { return _msg.c_str(); }
+
+private:
+    ss::sstring _msg;
+};
+
+/**
+ * Class representing an XID.
+ *
+ * The binary representation of the id has 12 bytes. String representation is
+ * using base32 encoding (w/o padding) for better space efficiency when stored
+ * in that form (20 bytes).
+ *
+ * Based on (https://github.com/rs/xid)
+ */
+class xid {
+public:
+    // number of bytes in an underlying data structure
+    static constexpr size_t size = 12;
+    // size of xid string representation
+    static constexpr size_t str_size = 20;
+
+    using data_t = std::array<uint8_t, size>;
+
+    explicit constexpr xid(data_t data)
+      : _data(data) {}
+
+    // default constructor to make the xid type lexically castable from string
+    constexpr xid() = default;
```

Comment:
i can't remember the rule. will _data be value-initialized here with zeros (or junk)? i'm assuming we want this to have a deterministic value?

Reply:
added default initialisation